### ISSUE TITLE
Application crashes at startup (v10.3.3.0) due to date parsing error

### ISSUE THREAD

**Issue by thomasburguiere** [EDDiscovery/EDDiscovery#2303]
stack trace

```
An unhandled fatal exception has occurred.

==== BEGIN ====
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.FormatException: String was not recognized as a valid DateTime.
   at System.DateTimeParse.Parse(String s, DateTimeFormatInfo dtfi, DateTimeStyles styles)
   at EliteDangerousCore.JournalEntry..ctor(JObject jo, JournalTypeEnum jtype) in C:\Code\EDDiscovery\EliteDangerous\EliteDangerous\JournalEntry.cs:line 419
   at EliteDangerousCore.JournalEvents.JournalFSSSignalDiscovered..ctor(JObject evt) in C:\Code\EDDiscovery\EliteDangerous\JournalEvents\JournalDiscoveryScans.cs:line 123
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(Type type, Object[] args)
   at EliteDangerousCore.JournalEntry.CreateJournalEntry(JObject jo) in C:\Code\EDDiscovery\EliteDangerous\EliteDangerous\JournalEntry.cs:line 1166
   at EliteDangerousCore.JournalEntry.CreateJournalEntry(DataRow dr) in C:\Code\EDDiscovery\EliteDangerous\EliteDangerous\JournalEntry.cs:line 425
   at EliteDangerousCore.JournalEntry.GetAll(Int32 commander, Nullable`1 after, Nullable`1 before, JournalTypeEnum[] ids, Nullable`1 allidsafter) in C:\Code\EDDiscovery\EliteDangerous\EliteDangerous\JournalEntry.cs:line 859
   at EliteDangerousCore.EDSM.EDSMLogFetcher.Process(List`1 edsmlogs, DateTime logstarttime, DateTime logendtime) in C:\Code\EDDiscovery\EliteDangerous\EDSM\EDSMLogFetcher.cs:line 173
   at EliteDangerousCore.EDSM.EDSMLogFetcher.FetcherThreadProc() in C:\Code\EDDiscovery\EliteDangerous\EDSM\EDSMLogFetcher.cs:line 137
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
===== END =====
```

**Comment by klightspeed** [comment]
That would suggest an entry saved to the database is corrupt.

**Comment by thomasburguiere** [comment]
Yes, seems so, any idea how I can find that entry, and any way I can delete it ?

**Comment by klightspeed** [comment]
Did it only recently start occurring?  Does the trace log (in `%LOCALAPPDATA%\EDDiscovery\Log`) mention any journal errors?

**Comment by thomasburguiere** [comment]
Yes, started today, even though I use eddiscovery since months now.

I couldn’t find the EDDiscovery folder you mention in my LOCALAPPDATA, I’m using the portable version, could this be the reason ?

Thanks :)

**Comment by klightspeed** [comment]
Try `Data\Log` in the directory you installed EDDiscovery.

**Comment by klightspeed** [comment]
That log unfortunately didn't give what event was in error

Rebuilding the system DB would hsve postponed the error until the rebuild was complete.

If you can grsb a log from shortly before you noticed it crashing, it should have an error from when EDDiscovery was originally processing the entry.

**Comment by robbyxp1** [comment]
We should protect
EventTimeUTC = DateTime.Parse(jo.Value<string>("timestamp"), CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal);
in the jo reader.. try parse..

**Comment by robbyxp1** [comment]
Or maybe in the CreateJournalEntry's (multiple) we protect against a bad entry, and return null if it is, and in the users of this accept that null can be returned and not store it.. that way a bad entry means it does not get put in, but does not crash the reader completely.

**Comment by thomasburguiere** [comment]
Here's the first log file where the error occurred
[Trace_2019-01-04-22-20-23.0.log](https://github.com/EDDiscovery/EDDiscovery/files/2729434/Trace_2019-01-04-22-20-23.0.log)

**Comment by robbyxp1** [comment]
Don't suppose you could find the journal file from frontier where that occurred..

Journal files are in your c:\ users\ saved games\frontier\etc folder..

You can tell the one by matching up the date time of the event (2019-01-04 5:44) against the filename.

if you could find the FSSSignalDiscovered we could find out if it was Frontier who was writing a bad record.

**Comment by robbyxp1** [comment]
Sorry, 23:05:41 I meant. 

folder is
c:\users\<user name>\saved games\frontier developments\elite dangerous>

**Comment by thomasburguiere** [comment]
Here's the journal file from the same day: 
[Journal.190104222054.01.log](https://github.com/EDDiscovery/EDDiscovery/files/2729443/Journal.190104222054.01.log)

**Comment by robbyxp1** [comment]
That log goes to 22:41 UTC, error happened just after 23:05:05 utc, have you got another one after that?

**Comment by thomasburguiere** [comment]
My bad, was looking into the wrong journal file, theres the one with the events of 2019-01-04-23:05:

[Journal.190104234128.01.log](https://github.com/EDDiscovery/EDDiscovery/files/2729453/Journal.190104234128.01.log)

But I dont see any FSSSignalDiscovered  event around that time in this one :/

**Comment by robbyxp1** [comment]
My plan, this morning, and i'm currently debugging it by putting bad entries into the DB, is to protect against this.  I'll release a new EDD today but mark it pre-release (so everyone does not get it signalled).  If you can try it later, that would be good

I'll let you know.

**Comment by robbyxp1** [comment]
10.3.4 marked as pre-release, try that, it coped with a bad entry in the time date field of the JSON.

That bit of code was very old.

**Comment by klightspeed** [comment]
If you open `EDDUser.sqlite` using e.g. SQLiteBrowser, and execute the following SQL command, does it return anything:

```
SELECT * FROM JournalEntries WHERE EventData NOT LIKE '%"timestamp": "____-__-__T__:__:__Z"%'
```

**Comment by thomasburguiere** [comment]
No it doesnt, but as I mentioned earlier, I reset the DB this morning.

**Comment by robbyxp1** [comment]
Since you can't retest, we can presume, since i replicated it, we have fixed the issue for 10.4.  Reopen if you see this again.

**Comment by thomasburguiere** [comment]
Yes, thanks a lot :)
